### DISCUSSION TITLE
Coaxial rotor

### DISCUSSION THREAD

**Discussion by mmajerTUD** [petrobras/ross#986] [Q&A]
Dear developers,
my colleague and I are trying ROSS for the first time to model and simulate the dynamic response of our high-speed rotor. 
With reference to the model attached (obtained with an in-house software), we have tried to reproduce it using ROSS without much success. 
The main shaft is supported on ball bearings and runs within a coaxial, non-rotating sleeve (the other body). The non-rotating sleeve's outer surface acts as a squeeze film damper. 
With ROSS, we tried to use the class Coaxial rotor but we get a wrong result (see the other picture attached, with the ROSS model obtained).
I also attach the code we used for your reference. 
Is there something I am doing wrong or is my case not possible to be modelled with ROSS?
![Model](https://github.com/petrobras/ross/assets/71271129/ecf95ff8-3678-408c-b328-34899544a72b)
![ROSS Model](https://github.com/petrobras/ross/assets/71271129/a1a61613-94bd-410c-b144-f8068e360fdf)

```
import numpy as np
import ross as rs

steel = rs.materials.steel

# 1. Main shaft discretization
L =   [0.005, 0.007, 0.007, 0.001, 0.009, 0.00539, 0.00259, 0.00666, 0.00192, 0.00683, 0.00842, 0.00436, 0.00837, 0.00231, 0.00629, 0.00198, 0.00065, 0.0007, 0.00065, 0.01817, 0.0036, 0.00796]
i_d = [    0,     0,     0,     0,     0,       0,       0,       0,       0,       0,       0,       0,       0,       0,       0,       0,       0,      0,       0,       0,      0,       0]
o_d = [0.008, 0.008, 0.008, 0.0096, 0.0096,   0.0096,   0.0096,   0.0096,   0.0096,   0.0096,   0.0096,   0.0096,   0.0096,    0.01,    0.01,    0.01,    0.01,    0.01,   0.01,   0.012,  0.008,   0.008]
N = len(L)
shaft_elements = [
    rs.ShaftElement(
        L=L[i],
        idl=i_d[i],
        odl=o_d[i],
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    for i in range(N)
]

main_shaft = shaft_elements

# 1.1 Check main shaft
R1 = rs.Rotor(main_shaft)
R1.plot_rotor().show()

# 2. Cartridge outer ring discretization
L =   [0.00539, 0.00259, 0.00666, 0.00192, 0.00683, 0.00842, 0.00436, 0.00837, 0.00231, 0.00629, 0.00198]
i_d = [0.02308, 0.02308, 0.02187, 0.02153, 0.02153, 0.02153, 0.02153, 0.02153, 0.02153, 0.02187, 0.02308]
o_d = [0.0273, 0.0273, 0.029, 0.029, 0.029, 0.02724, 0.02724, 0.02724, 0.029, 0.029, 0.029]
N = len(L)
shaft_elements = [
    rs.ShaftElement(
        L=L[i],
        idl=i_d[i],
        odl=o_d[i],
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    for i in range(N)
]

outer_ring = shaft_elements

# 2.1 Check cartridge
R2 = rs.Rotor(outer_ring)
R2.plot_rotor().show()

# 3 Assembly main shaft + cartridge
shaft = [main_shaft, outer_ring]

stf=1e6
brg1 = rs.BearingElement(6,n_link=24,kxx=stf,kyy=stf,cxx=0)
brg2 = rs.BearingElement(14,n_link=32,kxx=stf,kyy=stf,cxx=0)
bearings = [brg1, brg2]

disks = []

# 3.1 Check assembly main shaft + cartridge
A1 = rs.CoAxialRotor(shaft, disks, bearings)
A1.plot_rotor().show()

# 4 Add fitted flange on main shaft
L =   [0.005, 0.007]
i_d = [0.008, 0.008]
o_d = [0.022, 0.0114]
N = [0, 1]
shaft_elements = [
    rs.ShaftElement(
        L=L[i],
        n=N[i],
        idl=i_d[i],
        odl=o_d[i],
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    for i in range(len(N))
]

# create variable shaft composed by main shaft + flange
shaft = main_shaft
[shaft.append(shaft_elements[i]) for i in range(len(shaft_elements))]

# 4.1 Check main shaft + flange
R3 = rs.Rotor(shaft)
R3.plot_rotor().show()

# 5. Assembly shaft + cartridge
assembly = [shaft, outer_ring]

# 5.1 Check assembly shaft + cartridge
A2 = rs.CoAxialRotor(assembly, disks, bearings)
A2.plot_rotor().show()
```

**Comment by raphaeltimbo** [comment]
Hi @mmajerTUD !
Unfortunately, ROSS support for coaxial rotors is limited.
When you assemble the model you will be able to calculate natural frequencies, mode shapes etc. just fine.
But things like plotting the rotors are not working properly. 
We would have to dedicate sometime to implement those, but in Petrobras we don't simulate many coaxial rotors, so I don't think we will be able to prioritize this part of the code right now. 
If we prioritize this in the future I will let you know here in this 'discussion'.

**Reply by mmajerTUD** [reply]
Hi Raphael, thanks for your feedback! 
I understand your reasons, and I hope to hear more from you soon on this!
One question related to ROSS calculations on coaxial rotors; is it possible to assign different rotational speeds to the coaxial rotors? 
In the coaxial rotor class, I have seen a generic rotational speed attribute, does it mean I can only simulate situations where both rotors spin at the same speed?

**Reply by raphaeltimbo** [reply]
Hi @mmajerTUD! 
Sadly, right now you can only set one speed. 
In the past we've discussed the idea of a more general `MultiRotor` class that could change things #755. 
Unfortunatly we did not find the time to implement that one.

**Comment by JW97** [comment]
Has there been any further work on the topic of co-axial rotors?
We're working on an application where a rotor is rotating around a centre stationary shaft that is cantilevered on one side

**Reply by mmajerTUD** [reply]
Hi, a colleague of mine has indeed made a modification to the source code, but that code lives on a private fork. 
We have just recently discussed that we should pick this work back up soon, possibly in June. 
We can get back to you with an update later on on this thread and share the status of our current developments.
It would be nice to unify the efforts when developing this new capability.

**Reply by mmajerTUD** [reply]
Hi, just a quick update on this topic. 
We have not had the time to look into this in June, I am actively looking for somebody to support this effort as I am currently too busy myself to work on it first-hand.
Did you make any progress in the meantime?

**Reply by JW97** [reply]
First of all, thank you for keeping this issue in your mind. It's great to see the project being supported/upgraded over time!

We ultimately paused rotodynamics analysis for this project as a result of the tools limitation.
We would however like to revisit this, as the setup isn't uncommon for us.
